cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java

```diff
@@ -1058,4 +1300,50 @@ public boolean apply(Privilege input) {
     Assert.assertTrue(Sets.filter(authorizer.listPrivileges(ALICE), entityFilter).isEmpty());
     Assert.assertTrue(Sets.filter(authorizer.listPrivileges(BOB), entityFilter).isEmpty());
   }
+
+  private void verifyDatasetIsEmpty(NamespaceId namespaceId, String datasetName) throws Exception {
+    DataSetManager<KeyValueTable> outTableManager = getDataset(namespaceId.toId(), datasetName);
+    KeyValueTable outputTable = outTableManager.get();
+    CloseableIterator<KeyValue<byte[], byte[]>> scanner = outputTable.scan(null, null);
+    Assert.assertFalse(scanner.hasNext());
+    scanner.close();
+  }
+
+  private void verifyProgramFailure(Map<String, String> programArgs, ProgramManager programManager) throws
+    ExecutionException, InterruptedException, TimeoutException {
+    programManager.start(programArgs);
+    programManager.waitForFinish(5, TimeUnit.MINUTES);
+
+    // verify program history just have failures
+    List<RunRecord> history = programManager.getHistory();
+    for (final RunRecord runRecord : history) {
+      // the program status might not have stamped as failed instantly after the program finishes so wait for
+      // sometime for it to be marked as failed.
+      try {
+        Tasks.waitFor(ProgramRunStatus.FAILED, new Callable<ProgramRunStatus>() {
+          @Override
+          public ProgramRunStatus call() throws Exception {
+            return runRecord.getStatus();
+          }
+        }, 60, TimeUnit.SECONDS);
+      } catch (TimeoutException e) {
+        Assert.fail(String.format("Found program status %s while expecting it to be %s",
```

Comment:
I don't think we need this.
Tasks.waitFor should throw a meaningful message itself.